pyccel/ast/cudaext.py

```diff
@@ -9,14 +9,124 @@
 Provides CUDA functionality for code generation.
 """
 from .internals      import PyccelFunction
+from .literals       import Nil
 
 from .datatypes      import VoidType
 from .core           import Module, PyccelFunctionDef
+from .numpyext       import process_dtype, process_shape
+from .cudatypes      import CudaArrayType
+
+
 
 __all__ = (
     'CudaSynchronize',
+    'CudaNewarray',
+    'CudaFull',
+    'CudaHostEmpty'
 )
 
+class CudaNewarray(PyccelFunction):
+    """
+    Superclass for nodes representing Cuda array allocation functions.
+
+    Class from which all nodes representing a Cuda function which implies a call
+    to `Allocate` should inherit.
+
+    Parameters
+    ----------
+    *args : tuple of TypedAstNode
+        The arguments of the superclass PyccelFunction.
+
+    class_type : NumpyNDArrayType
+        The type of the new array.
+
+    init_dtype : PythonType, PyccelFunctionDef, LiteralString, str
+        The actual dtype passed to the Cuda function.
+
+    memory_location : str
+        The memory location of the new array ('host' or 'device').
+    """
+    __slots__ = ('_class_type', '_init_dtype', '_memory_location')
+    name = 'newarray'
+
+    def __init__(self, *args ,class_type, init_dtype, memory_location):
+        self._class_type = class_type
+        self._init_dtype = init_dtype
+        self._memory_location = memory_location
```

Comment:
Is the memory location not inside the class type already?
```suggestion
    def __init__(self, *args, class_type, init_dtype, memory_location):
        self._class_type = class_type
        self._init_dtype = init_dtype
        self._memory_location = memory_location
```